File: resources/tail-recursive/TailRecursiveTest.java
public class TailRecursiveTest {
    static int sumCommon(int n) {
        if (n == 0) return n;
        return sumCommon(n-1) + n;
    }

    static int sumTail(int n, int sum) {
        if (n == 0) return sum;
        return sumTail(n-1, sum+n);
    }

    public static void main(String[] args) {
        System.out.println(sumCommon(10));
        int sum = 0;
        System.out.println(sumTail(10, sum));
    }

}

File: resources/tail-recursive/tail-recursive-test.cpp
#include <iostream>
using namespace std;

int sumCommon(int n) {
    if (n <= 1) return n;
    return sumCommon(n-1) + n;
}

int sumTailRecursive(int n, int sum) {
    if (n <= 1) return sum+n;
    return sumTailRecursive(n-1, sum+n);
}

int main() {
    cout << sumCommon(10) << endl;
    int sum = 0;
    cout << sumTailRecursive(10, sum) << endl;
    return 0;
}

